compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -360,6 +360,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         false
     }
 
+    crate fn hir_id_sole_block_element(
+        &self,
+        hir_id: hir::HirId,
+    ) -> Option<&'tcx rustc_hir::Expr<'tcx>> {
+        let node: Option<Node<'_>> = self.tcx.hir().find(hir_id);
+        match node {
+            Some(Node::Expr(rustc_hir::Expr {
+                kind: rustc_hir::ExprKind::Block(block, ..),
+                ..
+            })) if block.stmts.len() == 0 => block.expr,
```

Comment:
IIUC this will only catch blocks with no statements. This seems like an unnecessary restriction to me, am I missing something?
Consider the following code:
```rust
fn main() {
    let a = &1;
    let b = &2;
    let val = if true {
        a + 1
    } else {
        let c = 0;
        b
    };
    println!("{}", b);
}
```

Where the block *has* statements, but still gives the incorrect suggestion:
```
help: consider dereferencing the borrow
  |
6 |     } else *{
7 |         let c = 0;
8 |         b
9 |     };
```

(Maybe add this as a test?)

Reply:
Yes, you are correct. I just wrongly assumed it didn't work properly only without statements in the else block. I have added a test to make sure this is correct as well.